### DISCUSSION TITLE
Is there any way to set BIGSERIAL, or just SERIAL for column that is not a PK?

### DISCUSSION THREAD

**Discussion by khoroshevskyi** [sqlalchemy/sqlalchemy#9790] [Usage Questions] [answered]
I really like a new way of declaring sachems and tables in sqlalchemy, but I ran into a few issues that I can't find in the documentation.

While using simple SQL, I am creating table with next values: 

```sql
CREATE TABLE projects (
    id BIGSERIAL NOT NULL,
    namespace TEXT NOT NULL,
    name TEXT NOT NULL,
    ...
    submission_date timestamp NOT NULL,
    last_update_date timestamp NOT NULL,
    CONSTRAINT id PRIMARY KEY (namespace, name, tag)
);
```

1) Is there any way, id can be serializable without setting it as PK?
2) Is there any way to create constraint PK the way I did above? 

*I am using sqlalchemy2.0  with new way of initializing Declarative Mapping

Thank you in advance for your help.

**Comment by zzzeek** [comment, marked as answer]
we dont have a SERIAL/BIGSERIAL type in the PG dialect directly but you can create one, as below.   The column requires a FetchedValue so that the ORM does not try to insert NULL into it.  It includes the [PrimaryKeyConstraint](https://docs.sqlalchemy.org/en/20/core/constraints.html#sqlalchemy.schema.PrimaryKeyConstraint) syntax you asked for.


```py

from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import PrimaryKeyConstraint, FetchedValue
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped, Session
from sqlalchemy.orm import mapped_column
from sqlalchemy.types import BigInteger


class BIGSERIAL(BigInteger):
    pass


@compiles(BIGSERIAL, "postgresql")
def compile_bigserial_pg(type_, compiler, **kw):
    return "BIGSERIAL"


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(BIGSERIAL, server_default=FetchedValue())

    namespace: Mapped[str]
    name: Mapped[str]
    tag: Mapped[str]

    __table_args__ = (
        PrimaryKeyConstraint("namespace", "name", "tag", name="id"),
    )


e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)
s.add(A(namespace="np", name="nm", tag="tg"))
s.commit()

```

**Reply by CaselIT** [reply]
I've updated the example to make BIGSERIAL extend from BigInteger instated of Integer

**Reply by khoroshevskyi** [reply]
Great! Thank you very much for your quick assistance!
